src/Distributed/distributed_fft_based_poisson_solver.jl

```diff
@@ -1,134 +1,339 @@
+        end
+    end
+
+    # Build transforms
+    transforms = Tuple(infer_transform(effective_topology[d]) for d in Tuple(input_permutation))
+    plan = PencilFFTPlan(permuted_size, transforms, processors_per_dimension, communicator)
 
-    λx = solver.eigenvalues[1]
-    λy = solver.eigenvalues[2]
-    λz = solver.eigenvalues[3]
+    # Allocate memory for in-place FFT + transpositions
+    transposition_storage = PencilFFTs.allocate_input(plan)
+
+    # Store a view of the right hand side that "appears" to have the permutation (x, y, z).
+    permuted_right_hand_side = first(transposition_storage)
+    inverse_permutation = Tuple(findfirst(p -> p == d, Tuple(input_permutation)) for d in 1:3)
+    unpermuted_right_hand_side = PermutedDimsArray(permuted_right_hand_side, inverse_permutation)
+
+    if using_tridiagonal_vertical_solver
+        lower_diagonal, diagonal, upper_diagonal = compute_batched_tridiagonals(grid, λx, λy)
+        tridiagonal_vertical_solver = BatchedTridiagonalSolver(grid; lower_diagonal, diagonal, upper_diagonal)
+        tridiagonal_storage = zeros(eltype(first(transposition_storage)), child_architecture(local_grid), size(local_grid)...)
+    else
+        tridiagonal_vertical_solver = nothing
+        tridiagonal_storage = nothing
+    end
+    
+    # Permute the λ appropriately
+    permuted_eigenvalues = Tuple(unpermuted_eigenvalues[d] for d in Tuple(input_permutation))
+    eigenvalues = PencilFFTs.localgrid(last(transposition_storage), permuted_eigenvalues)
 
-    # Apply forward transforms to b = first(solver.storage).
-    solver.plan * solver.storage
+    return DistributedFFTBasedPoissonSolver(plan, global_grid, local_grid, eigenvalues, transposition_storage, input_permutation,
+                                            unpermuted_right_hand_side, tridiagonal_vertical_solver, tridiagonal_storage)
+end
 
-    # Solve the discrete Poisson equation in wavenumber space
-    # for x̂. We solve for x̂ in place, reusing b̂.
-    x̂ = b̂ = last(solver.storage)
+"""
+    solve_transformed_poission!(solver)
 
-    @. x̂ = - b̂ / (λx + λy + λz)
+Solve the discrete Poisson equation after transforming into the three-dimensional
+eigenfunction space of the second-order discrete Poisson operator.
+We solve the Poisson equation "in-place", re-using the transformed
+right-hand-side, `b̂ = last(solver.transposition_storage)` to store the solution `x̂`.
+"""
+function solve_transformed_poission_equation!(solver)
+    x̂ = b̂ = last(solver.transposition_storage)
+    λ = solver.eigenvalues
+    @. x̂ = - b̂ / (λ[1] + λ[2] + λ[3])
 
     # Set the zeroth wavenumber and volume mean, which are undetermined
     # in the Poisson equation, to zero.
-    if MPI.Comm_rank(solver.communicator) == 0 #MPI.Comm_size() - 1
-        x̂[1, 1, 1] = 0
+    multi_arch = architecture(solver.local_grid)
+
+    if MPI.Comm_rank(multi_arch.communicator) == 0
+        # This is an assumption: we *hope* PencilArrays allocates in this way
```

Comment:
This should usually be true. You can verify it using `range_local`:

```julia
rng = PencilArrays.range_local(x̂)  # returns local range (i1:i2, j1:j2, k1:k2) in the global grid
@assert all(first.(rng) .== 1)  # check that i1 == j1 == k1 == 1
```

Reply:
Nice, thank you!

Reply:
Is there a way to change the condition to always be true? Ie, can we diagnose which rank has (1, 1, 1) (cheaply?)

Reply:
I guess you provided precisely what we need for that:

```julia
ii, jj, kk = PencilArrays.range_local(x̂)
if (ii[1], jj[1], kk[1]) === (1, 1, 1)
    # zero it out
end
```